internal/binding/local.go

```diff
@@ -412,127 +417,196 @@ func (lb *LocalBind) createATE(_ context.Context, ate *opb.Device, portMgr *ateP
-	}, nil
+	return nil
 }
 
-func (p *chanPort) Close() {
-	p.txQueue.Close()
+// CreateHandle implements  magna's API for creating handles.
+func (pm *portMgr) CreateHandle(name string) (common.Port, error) {
+	port, ok := pm.ports[name]
+	if !ok {
+		return nil, fmt.Errorf("port %v not found", name)
+	}
+	return port.newHandle(), nil
 }
 
-func (p *chanPort) LinkType() layers.LinkType {
-	return layers.LinkTypeEthernet
+func (pm *portMgr) dutManager(dutID string) *dutManager {
+	return &dutManager{
+		mgr:   pm,
+		dutID: dutID,
+	}
 }
 
-func (p *chanPort) SetBPFFilter(string) error {
+func (pm *portMgr) linkPorts(a, b string) error {
+	aPort := pm.ports[a]
+	bPort := pm.ports[b]
+
+	if aPort == nil || bPort == nil {
+		return fmt.Errorf("ports do not exist: a %v b %v", a, b)
+	}
+	aPort.rxQueue = bPort.txQueue
+	bPort.rxQueue = aPort.txQueue
+
 	return nil
 }
 
-type portMgr struct {
-	dut *dutPortMgr
-	ate *atePortMgr
+// dutManager handles mapping from the SAI API hardware lane to Ondatra port ID.
+type dutManager struct {
+	mgr   *portMgr
+	dutID string
 }
 
-// dutPortMgr implements lemming's port creator interface.
-// Note: this interface is different from magna, to avoid making the two project depending on eachother.
-type dutPortMgr struct {
-	ports map[string]*chanPort
-	// TODO: lastCreatedPort is a hack because we don't currently model hardware ports correctly in lemming.
-	lastCreatedPort *chanPort
+// CreatePort impplements lemming's API for creating port.
+func (dm *dutManager) CreatePort(name string) (fwdcontext.Port, error) {
+	port, ok := dm.mgr.ports[dm.mgr.dutLaneToPort[dm.dutID][name]]
+	if !ok {
+		return nil, fmt.Errorf("port %v not found", name)
+	}
+	return port.newHandle(), nil
 }
 
-func (mgr *dutPortMgr) CreatePort(name string) (fwdcontext.Port, error) {
-	q, err := queue.NewUnbounded(fmt.Sprintf("%s-tx", name))
-	if err != nil {
-		return nil, err
-	}
-	p := &chanPort{
-		txQueue: q,
-	}
-	mgr.lastCreatedPort = p
-	q.Run()
-	return p, nil
+type chanPort struct {
+	mu sync.RWMutex
+
+	txQueue *queue.Queue
+	rxQueue *queue.Queue
+	handles []*portHandle
+}
+
+func (p *chanPort) newHandle() *portHandle {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	h := &portHandle{
+		rx: make(chan []byte, 1024),
+		tx: p.txQueue,
+	}
+	p.handles = append(p.handles, h)
+	i := len(p.handles) - 1
+	h.closeFn = func() {
+		p.mu.Lock()
+		defer p.mu.Unlock()
+		close(h.rx)
+		p.handles = append(p.handles[:i], p.handles[i+1:]...)
+	}
+	return h
+}
+
+func (p *chanPort) run() {
+	go func() {
+		for {
+			if p.rxQueue == nil {
+				continue
+			}
+			packet := (<-p.rxQueue.Receive()).([]byte)
+			p.mu.RLock()
+			for _, h := range p.handles {
+				h.rx <- packet
+			}
+			p.mu.RUnlock()
+		}
+	}()
 }
 
-// atePortMgr implement magna's port creator interface.
-// Note: this interface is different from lemming, to avoid making the two project depending on eachother.
-type atePortMgr struct {
-	ports map[string]*chanPort
+type portHandle struct {
```

Comment:
Can we have comment on this struct? look to me this is only for packet capture?

Reply:
done